pkg/azure/keyvault.go

```diff
@@ -0,0 +1,310 @@
+	if err != nil {
+		return nil, fmt.Errorf("getting az credentials: %w", err)
+	}
+
+	client, err := armkeyvault.NewVaultsClient(subscriptionId, cred, nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating client: %w", err)
+	}
+
+	pager := client.NewListByResourceGroupPager(resourceGroup, nil)
+	var akvs []Akv
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("getting next page: %w", err)
+		}
+
+		for _, kv := range page.Value {
+
+			if kv == nil {
+				return nil, fmt.Errorf("nil keyvault")
+			}
+			id, err := arm.ParseResourceID(*kv.ID)
+			if err != nil {
+				return nil, fmt.Errorf("parsing resource id: %w", err)
+			}
+			newAkv := LoadAkv(*id)
+			newAkv.Uri = *kv.Properties.VaultURI
+
+			akvs = append(akvs, *newAkv)
+			lgr.Info("keyvault", "name", kv.Name)
+		}
+	}
+	return akvs, nil
+}
+
+func NewAkv(ctx context.Context, tenantId, subscriptionId, resourceGroup, name, location string) (*Akv, error) {
+	name = truncate(name, 24)
+
+	lgr := logger.FromContext(ctx).With("name", name, "resourceGroup", resourceGroup, "location", location, "subscriptionId", subscriptionId)
+	ctx = logger.WithContext(ctx, lgr)
+	lgr.Info("starting to create akv")
+	defer lgr.Info("finished creating akv")
+
+	cred, err := getCred()
+	if err != nil {
+		return nil, fmt.Errorf("getting az credentials: %w", err)
+	}
+
+	factory, err := armkeyvault.NewClientFactory(subscriptionId, cred, nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating client factory: %w", err)
+	}
+
+	clientObjectId, err := getObjectId(ctx, cred)
+	if err != nil {
+		return nil, fmt.Errorf("getting client object id: %w", err)
+	}
+
+	v := &armkeyvault.VaultCreateOrUpdateParameters{
+		Location: to.Ptr(location),
+		Properties: &armkeyvault.VaultProperties{
+			AccessPolicies: []*armkeyvault.AccessPolicyEntry{
+				{
+					ObjectID: to.Ptr(clientObjectId),
+					Permissions: &armkeyvault.Permissions{
+						Secrets: []*armkeyvault.SecretPermissions{
+							to.Ptr(armkeyvault.SecretPermissionsGet),
+							to.Ptr(armkeyvault.SecretPermissionsSet),
+						},
+					},
+					TenantID:      to.Ptr(tenantId),
+					ApplicationID: nil,
+				},
+			},
+			TenantID: to.Ptr(tenantId),
+			SKU: &armkeyvault.SKU{
+				Name: to.Ptr(armkeyvault.SKUNameStandard),
+			},
+		},
+	}
+	poller, err := factory.NewVaultsClient().BeginCreateOrUpdate(ctx, resourceGroup, name, *v, nil)
+	if err != nil {
+		return nil, fmt.Errorf("starting to create vault: %w", err)
+	}
+
+	result, err := poller.PollUntilDone(ctx, nil)
+	if err != nil {
+		return nil, fmt.Errorf("waiting for vault creation to complete: %w", err)
+	}
+
+	return &Akv{
+		Uri:            *result.Properties.VaultURI,
+		Id:             *result.ID,
+		ResourceGroup:  resourceGroup,
+		Name:           *result.Name,
+		SubscriptionId: subscriptionId,
+		TenantId:       tenantId,
+	}, nil
+}
+
+func (a *Akv) PutSecretIfNewValue(ctx context.Context, name, value string) error {
```

Comment:
why include "if new value" in the function name?

Reply:
oh that's to avoid making a new version every time we put the secret, even if the value isn't updated. Otherwise we make unnecessary version